### PR TITLE
Re-enable joining from old snapshot test

### PR DESCRIPTION
**Pull request by jumaffre**
The `run_join_old_snapshot` seems to have been disabled for a while now so this PR re-enables it.

This also contains a fix so that nodes started with no snapshots (i.e. need to replay the entire historical ledger on join) are also subject to the same check.

We now also try to join/recover from a snapshot as much as possible in the Python infra.

Note: the justification for this is outlined in the original PR https://github.com/microsoft/CCF/pull/2422.

### CODE REVIEW

**Review by wintersteiger — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 - Nodes no longer crash at start-up if the ledger in the read-only ledger directories (`ledger.read_only_directories`) is ahead of the ledger in the main ledger directory (`ledger.directory`) (#3597).
 
+### Fixed
+
+- Fixed an issue where new node started without a snapshot would be able to join from a node that started with a snapshot (#3573).
```
Comment (achamayou) [inline CHANGELOG.md:27]: Missing "not"?

**Review by achamayou — APPROVED** [review]

**Review by raghavthind2005 — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -97,6 +97,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 - Nodes no longer crash at start-up if the ledger in the read-only ledger directories (`ledger.read_only_directories`) is ahead of the ledger in the main ledger directory (`ledger.directory`) (#3597).
 
+### Fixed
+
```
Comment (achamayou) [inline CHANGELOG.md:101]: Needs moving up to Unreleased.

**Review by achamayou — APPROVED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (jumaffre) [reply CHANGELOG.md:101]: Well spotted!

**Pull request merged**